### PR TITLE
python3Packages.cylp: init at 0.91.4

### PR DESCRIPTION
**Pull request by rmcgibbo**
###### Motivation for this change
Adds the package requested here: https://github.com/NixOS/nixpkgs/issues/112353, and adds it as a dependency of CVXPY, so that it has access to another good solver.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by rmcgibbo** [comment]
Result of `nixpkgs-review pr 154026` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>18 packages built:</summary>
  <ul>
    <li>python38Packages.cvxpy</li>
    <li>python38Packages.cylp</li>
    <li>python38Packages.qiskit</li>
    <li>python38Packages.qiskit-aer</li>
    <li>python38Packages.qiskit-aqua</li>
    <li>python38Packages.qiskit-finance</li>
    <li>python38Packages.qiskit-ibmq-provider</li>
    <li>python38Packages.qiskit-ignis</li>
    <li>python38Packages.qiskit-optimization</li>
    <li>python39Packages.cvxpy</li>
    <li>python39Packages.cylp</li>
    <li>python39Packages.qiskit</li>
    <li>python39Packages.qiskit-aer</li>
    <li>python39Packages.qiskit-aqua</li>
    <li>python39Packages.qiskit-finance</li>
    <li>python39Packages.qiskit-ibmq-provider</li>
    <li>python39Packages.qiskit-ignis</li>
    <li>python39Packages.qiskit-optimization</li>
  </ul>
</details>

File: pkgs/development/python-modules/cvxpy/default.nix [hunk 1]
```diff
@@ -33,6 +34,7 @@ buildPythonPackage rec {
     osqp
     scipy
     scs
+    cylp
```
Comment (jonringer) [inline pkgs/development/python-modules/cvxpy/default.nix:37]: Is this an implicit dependency?

The build is fine on master. Generally optional dependencies are not included, as it increases minimal closure size.

Comment (rmcgibbo) [reply pkgs/development/python-modules/cvxpy/default.nix:37]: All of the solvers (osqp, scipy, scs, cvxopt included) here are optional.

Comment (rmcgibbo) [reply pkgs/development/python-modules/cvxpy/default.nix:37]: Coin-OR happens to be the _best_ open source solver that CVXPY supports as a backend, IMHO.

Comment (jonringer) [reply pkgs/development/python-modules/cvxpy/default.nix:37]: meh, I guess this should only get used in domain specific  scenarios.  And the closure is already massive

```
$ nix path-info -Sh $(nix-build -A python3Packages.cvxpy)
/nix/store/ay0fkmz55x3327p8xnwgvra3qigd2rcm-python3.9-cvxpy-1.1.18	 406.4M
```

File: pkgs/development/python-modules/cylp/default.nix [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    export CYLP_USE_CYTHON=1
+  '';
+
+  checkInputs = [ pytestCheckHook ];
+  pythonImportsCheck = [ "cylp" ];
+  pytestFlagsArray = [ "-s" ];
+
+  preCheck = ''
+    cd cylp/tests
+    rm __init__.py
```
Comment (drewrisinger) [inline pkgs/development/python-modules/cylp/default.nix:59]: Why? pytest shouldn't care about this (either exact directory or .

If decide to leave:
```suggestion
pytestFlagsArray = [ ... "cylp/tests"];
preCheck = ''
  rm cylp/tests/__init__.py
'';
```

File: pkgs/development/python-modules/cylp/default.nix [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  '';
+
+  # https://github.com/coin-or/CyLP/commit/e5bab3cca81c0ad9886bcd8127923a436bcdaf9b
```
Comment (drewrisinger) [inline pkgs/development/python-modules/cylp/default.nix:62]: How is this commit relevant to the disabled test file? I don't see anything about test_MIP at all.

File: pkgs/development/python-modules/cylp/default.nix [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  disabledTestPaths = [
+    "test_MIP.py"
```
Comment (drewrisinger) [inline pkgs/development/python-modules/cylp/default.nix:64]: It makes me feel a little uncomfortable disabling a test that is core functionality: from the description, "A Python interface to CLP, CBC, and CGL to solve LPs and MIPs"

File: pkgs/development/python-modules/cvxpy/default.nix (hunk 1 above)
Comment (drewrisinger) [inline pkgs/development/python-modules/cvxpy/default.nix:37]: Nitpick, but ``cylp`` should be in this list in alphabetical order (and in the arguments list)

File: pkgs/development/python-modules/cylp/default.nix [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  ];
+
+  preBuild = ''
+    # Leak some memory, reverting this commit. Otherwise there's a segfault
+    # that's being debugged in https://github.com/coin-or/CyLP/issues/105
+    substituteInPlace cylp/cy/CyClpSimplex.pyx --replace \
+      'del self.CppSelf' \
+      'pass'
+
+    # Rebuild all cython files
```
Comment (drewrisinger) [inline pkgs/development/python-modules/cylp/default.nix:49]: Isn't that implicit in the Nix build process? Comment doesn't seem useful.

File: pkgs/development/python-modules/cylp/default.nix [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+, fetchFromGitHub
+, numpy
+, scipy
+, python
+, pytestCheckHook
+, cython
+, cbc
+, bzip2
+, zlib
+, pkg-config
```
Comment (drewrisinger) [inline pkgs/development/python-modules/cylp/default.nix:12]: Personal preference is to sort these in some order: either in usage order below, or alphabetical.

**Review by drewrisinger — CHANGES_REQUESTED** [review]
* [x] Diff needs some small work. Some minor questions on testing, and a few nitpicks
* [ ] Squash commits?
* [ ] Haven't attempted build

File: pkgs/development/python-modules/cylp/default.nix (hunk 4 above)
Comment (rmcgibbo) [reply pkgs/development/python-modules/cylp/default.nix:64]: I agree, but unfortunately this particular test is _obviously_ broken because upstream removed a method without removing the calls to that method from inside the tests. I have verified that the functionality does work. Do you want me to add one of my own tests?

Comment (rmcgibbo) [reply pkgs/development/python-modules/cylp/default.nix:64]: I should have left more comments related to that URL, but the URL is relevant because this commit removed the branchAndBound function, and every test in that files calls the branchAndBound function.

Thus, unfortunately this particular test is _obviously_ broken because upstream due to the missing. I have verified that the functionality for solving MIPS does in fact work. Do you want me to add one of my own tests?

File: pkgs/development/python-modules/cylp/default.nix (hunk 3 above)
Comment (rmcgibbo) [reply pkgs/development/python-modules/cylp/default.nix:62]: See comment below

File: pkgs/development/python-modules/cylp/default.nix (hunk 2 above)
Comment (rmcgibbo) [reply pkgs/development/python-modules/cylp/default.nix:65]: I'm not entirely sure, but pytest gets confused with the local imports, since the tests are inside the source directory, if i don't do this.

File: pkgs/development/python-modules/cylp/default.nix (hunk 6 above)
Comment (rmcgibbo) [reply pkgs/development/python-modules/cylp/default.nix]: I sorted by usage order.

File: pkgs/development/python-modules/cylp/default.nix (hunk 5 above)
Comment (rmcgibbo) [reply pkgs/development/python-modules/cylp/default.nix]: I wrote a bit more in the comment about why this is a good idea. They ship codegen'd  C++ files on github, but for us its better to regenerate them.

File: pkgs/development/python-modules/cylp/default.nix (hunk 4 above)
Comment (AndersonTorres) [reply pkgs/development/python-modules/cylp/default.nix:70]: Was the upstream warned about this lack of updates?

You do not need to run your own tests here, in this particular case it is the upstream that is in error.

Comment (rmcgibbo) [reply pkgs/development/python-modules/cylp/default.nix:70]: Yes, the upstream is definitely aware. I may submit a patch to fix these tests to upstream when I have some free time as well.

File: pkgs/development/python-modules/cylp/default.nix (hunk 2 above)
Comment (drewrisinger) [reply pkgs/development/python-modules/cylp/default.nix:63]: If the package does something funky with local imports, I've had luck with either moving the tests to a separate directory, or changing pytest's import more ``--import-mode=append``. For future reference: https://docs.pytest.org/en/6.2.x/pythonpath.html#import-modes

**Review by drewrisinger — APPROVED** [review]
* [x] Diff LGTM. Thanks for addressing comments
* [x] Commits LGTM
* [x] Built cvxpy on x86_64-linux, didn't build downstream of that.

File: pkgs/development/python-modules/cylp/default.nix (hunk 4 above)
Comment (nessdoor) [reply pkgs/development/python-modules/cylp/default.nix:70]: Looks like upstream was just reminded of the issue in coin-or/CyLP#168.
It doesn't feel like they're going to fix this anytime soon unless someone provides them a new stimulus of some sort.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
please rebase

**Comment by rmcgibbo** [comment]
@Janik-Haag: I've abandoned this. Please feel free to go ahead and copy from my branch and make a new PR. I'll close this PR for now.
